### PR TITLE
Add Flip ranger operation to Roaring64Bitmap

### PR DESCRIPTION
**Pull request by simeonpilgrim**
We were wanting a Flip range operation, so copied the existing "could be improved" operation used in the 32 bit implementation.

Added a few helper functions to work around change from int for the high word, to byte[].

Added unit tests to cover the cases I could think of while developing.

As a side note question if there is a way to get an enumerable range from the ART then the algorithm can be transposed to stepping over the existing containers in the range, and injected the gaps, and altering the existing nodes. This would remove the polling/search of every node in the range.

### CODE REVIEW

File: RoaringBitmap/src/main/java/org/roaringbitmap/longlong/LongUtils.java [hunk 1]
```diff
@@ -125,4 +125,43 @@ private static int compareTo(byte[] buffer1, int offset1, int length1,
   public static long initWithFirst4Byte(int v) {
     return ((long) v) << 32;
   }
+
+  /**
+   * shift the long left by the container size amount so we can loop across containers by +1 steps
+   * @param num long being treated as unsigned long
+   * @return value shifted out of value space into container high part
+   */
+  public static long leftShiftHighPart(long num) {
```
Comment (weijietong) [inline RoaringBitmap/src/main/java/org/roaringbitmap/longlong/LongUtils.java:134]: Is the name right ? not `rightShiftHightPart` ?

File: RoaringBitmap/src/main/java/org/roaringbitmap/longlong/LongUtils.java [hunk 2]
```diff
@@ -125,4 +125,43 @@ private static int compareTo(byte[] buffer1, int offset1, int length1,
...
+    return num >>> 16;
+  }
+
+  /**
+   * shift the long by right the container size amount so we use the value after have done our steps
+   * @param num uint48 to be shift back into uint64
+   * @return value shifted out of container high part back into value space
+   */
+  public static long rightShiftHighPart(long num) {
```
Comment (weijietong) [inline RoaringBitmap/src/main/java/org/roaringbitmap/longlong/LongUtils.java:143]: the same as above

File: RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java [hunk 3]
```diff
@@ -298,6 +298,61 @@ public void andNot(final Roaring64Bitmap x2) {
...
 
+  /**
+   * Complements the bits in the given range, from rangeStart (inclusive) rangeEnd (exclusive). The
+   * given bitmap is unchanged.
+   *
+   * @param rangeStart inclusive beginning of range, in [0, 0xffffffffffffffff]
+   * @param rangeEnd exclusive ending of range, in [0, 0xffffffffffffffff + 1]
+   */
+  public void flip(final long rangeStart, final long rangeEnd) {
+    //rangeSanityCheck(rangeStart, rangeEnd);
```
Comment (weijietong) [inline RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java:309]: please remove the comment code

File: RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java [hunk 4]
```diff
@@ -298,6 +298,61 @@ public void andNot(final Roaring64Bitmap x2) {
     }
   }
 
+  /**
+   * Complements the bits in the given range, from rangeStart (inclusive) rangeEnd (exclusive). The
+   * given bitmap is unchanged.
+   *
+   * @param rangeStart inclusive beginning of range, in [0, 0xffffffffffffffff]
```
Comment (weijietong) [inline RoaringBitmap/src/main/java/org/roaringbitmap/longlong/Roaring64Bitmap.java:305]: please add some tests to the range edge cases.

File: RoaringBitmap/src/test/java/org/roaringbitmap/longlong/TestRoaring64Bitmap.java [hunk 5]
```diff
@@ -906,6 +906,158 @@ public void testAndNot_MultipleBucket() {
     assertEquals(123, left.select(0));
   }
 
+  @Test
+  public void testFlip_SameContainer() {
```
Comment (weijietong) [inline RoaringBitmap/src/test/java/org/roaringbitmap/longlong/TestRoaring64Bitmap.java:910]: please remove all the underline from the test method names

**Review by weijietong — COMMENTED** [review]
thanks for your contribution! please address the comments.

File: RoaringBitmap/src/main/java/org/roaringbitmap/longlong/LongUtils.java (hunk 1 above)
Comment (simeonpilgrim) [reply RoaringBitmap/src/main/java/org/roaringbitmap/longlong/LongUtils.java:134]: OMG, you are so correct. Thank you for spotting..

File: RoaringBitmap/src/test/java/org/roaringbitmap/longlong/TestRoaring64Bitmap.java (hunk 5 above)
Comment (simeonpilgrim) [reply RoaringBitmap/src/test/java/org/roaringbitmap/longlong/TestRoaring64Bitmap.java]: done

**Comment by weijietong** [comment]
LGTM +1
